hail/python/hailtop/batch_client/aioclient.py

```diff
@@ -453,89 +566,103 @@ def _create_job(self,
-                else:
-                    parent_ids.append(job._job_id)
+                    foreign_jobs.append(parent)
+                elif not 0 < job._spec.rel_job_id <= len(self._jobs):
+                    invalid_job_ids.append(parent)
             else:
-                foreign_batches.append(job)
+                assert isinstance(job, SubmittedJob)
+                if self._batch is None or job._batch != self._batch:
+                    foreign_jobs.append(parent)
 
         error_msg = []
-        if len(foreign_batches) != 0:
+        if len(foreign_jobs) != 0:
             error_msg.append(
                 'Found {} parents from another batch:\n{}'.format(
-                    str(len(foreign_batches)), "\n".join([str(j) for j in foreign_batches])
+                    str(len(foreign_jobs)), "\n".join([str(j._job) for j in foreign_jobs])
                 )
             )
         if len(invalid_job_ids) != 0:
             error_msg.append(
                 'Found {} parents with invalid job ids:\n{}'.format(
-                    str(len(invalid_job_ids)), "\n".join([str(j) for j in invalid_job_ids])
+                    str(len(invalid_job_ids)), "\n".join([str(j._job) for j in invalid_job_ids])
                 )
             )
         if error_msg:
             raise ValueError("\n".join(error_msg))
 
-        job_spec = {'always_run': always_run, 'job_id': self._job_idx, 'parent_ids': parent_ids, 'process': process}
-
-        if env:
-            job_spec['env'] = [{'name': k, 'value': v} for (k, v) in env.items()]
-        if port is not None:
-            job_spec['port'] = port
-        if resources:
-            job_spec['resources'] = resources
-        if secrets:
-            job_spec['secrets'] = secrets
-        if service_account:
-            job_spec['service_account'] = service_account
-        if timeout:
-            job_spec['timeout'] = timeout
-
-        if attributes:
-            job_spec['attributes'] = attributes
-        if input_files:
-            job_spec['input_files'] = [{"from": src, "to": dst} for (src, dst) in input_files]
-        if output_files:
-            job_spec['output_files'] = [{"from": src, "to": dst} for (src, dst) in output_files]
-        if cloudfuse:
-            job_spec['cloudfuse'] = [{"bucket": bucket, "mount_path": mount_path, "read_only": read_only}
-                                     for (bucket, mount_path, read_only) in cloudfuse]
-        if requester_pays_project:
-            job_spec['requester_pays_project'] = requester_pays_project
-        if mount_tokens:
-            job_spec['mount_tokens'] = mount_tokens
-        if network:
-            job_spec['network'] = network
-        if unconfined:
-            job_spec['unconfined'] = unconfined
-        if user_code:
-            job_spec['user_code'] = user_code
-
-        self._job_specs.append(job_spec)
-
-        j = Job.unsubmitted_job(self, self._job_idx)
+        job_spec = JobSpec(
```

Comment:
In the future, AFAICT, this change can happen independently of the open batches change and should happen in a separate PR for ease of review.